### PR TITLE
Added Web-GUI

### PR DESCRIPTION
**Pull request by shilch**
See #197 for To-dos
Added a Web-GUI (beta, I will add much more)
- Webserver is running on port 8000
- Socketserver is running on port 8001

bot

### CODE REVIEW

**Comment by jabbink** [comment]
Now that Spring Boot has been merged I think you could/should use that for providing the network access. That was the idea @apottere, right?

**Comment by Lenovotro** [comment]
Looks great!

**Comment by shilch** [comment]
Mmh, sorry I do not understand what Spring Boot is.
Also I do not know how to use the new access.
My Web-GUI uses a Socket Server which is the fastest connection that is possible. And it is easy to extend too

**Comment by iamthasasquatch** [comment]
I am getting when trying to build with this

e: C:\PogoDev\PokemonGoBotold\src\main\kotlin\ink\abb\pogo\scraper\gui\SocketSer
ver.kt: (102, 33): Unresolved reference. None of the following candidates is app
licable because of receiver type mismatch:
public fun Pokemon.getIvPercentage(): Int defined in ink.abb.pogo.scraper.util.p
okemon
:compileKotlin FAILED

**Comment by shilch** [comment]
Mmh, I do not get errors. Which Kotlin version do you use?

**Comment by iamthasasquatch** [comment]
anyway to have the map autozoom in right now its zoomed out to the entire world on starting. Also I feel the pokemon needs to be scaled way smaller right now its huge. Possibly 2 columns to show more on screen at same time

**Comment by RaeneDev** [comment]
Sorry for asking, but how can we access/view this?

**Comment by Forty-Tw0** [comment]
I presume it will be on localhost:8000

**Comment by Emuuung** [comment]
Is this being resolved?

**Comment by iamthasasquatch** [comment]
I edited the page and split a frame so on the left side can see pokemon without losing on screen information i am sure you could code it alot better than I did but here is an image of what I did.
frame

**Comment by shilch** [comment]
Yes that also looks good. But I will add more information to this page.

**Comment by iamthasasquatch** [comment]
I just like to see the map while scanning through my pokemon perhaps if you are adding inventory you can scale down the pokemon and split that frame for pokemon and then under it inventory. I am not very good at coding so mine is very sloppy and mostly just copy/paste of your code.

**Comment by Emuuung** [comment]
Can't test it cause the branch has conflicts :disappointed:

**Comment by shilch** [comment]
https://github.com/shilch/PokemonGoBot Use my fork

**Comment by Emuuung** [comment]
@shilch I clone your fork, build and run from source but http://localhost:8000/ isn't working.

**Comment by MASSV** [comment]
@jabbink springboot going to stay or was it causing too much problems? otherwise this seems to be pretty far along.

**Comment by jabbink** [comment]
@MASSV Spring Boot wasn't the problem. The problem was everything else done in that release, mainly reintroducing many already fixed bugs and removing already implemented features.

**Comment by leontada** [comment]
Guys seriously, this looks awesome!

**Comment by shilch** [comment]
I changed the UI. @iamthasasquatch now you can customize the gui (drag and resize windows)
I think this can be merged now. I will add some more soon (See #197 ).
img

**Comment by RobKellett** [comment]
Looks great, but as a minor complaint you should be using "Caught" instead of "Catched"

**Comment by shilch** [comment]
You are right, forgot the irregular verbs 😄

**Comment by Evuse** [comment]
@shilch what should i do to use your web interface? Can you provide a complete ZIP? Thanks :)

**Comment by iamthasasquatch** [comment]
I am getting a error when trying to do the new build.

e: C:\Dev\PokemonGoBot\src\main\kotlin\ink\abb\pogo\scraper\tasks\WalkToUnusedPo
kestop.kt: (75, 12): Unresolved reference: sendPokestops
:compileKotlin FAILED

**Comment by Evuse** [comment]
I'm using a VPS to run the script... i can't see anything on IP:8000 :(

**Comment by shilch** [comment]
@iamthasasquatch Marke sure that all files are up to date
@Evuse which project did you clone?

**Comment by pr0ves** [comment]
I added the item count, you can add it in the profile frame.

**Comment by shilch** [comment]
@pr0ves Thx, I will add this once I have enough time.

**Comment by iamthasasquatch** [comment]
running the latest clone and still getting an error and cant figure it out...

`e: C:\Dev\PokemonGoBot\src\main\kotlin\ink\abb\pogo\scraper\gui\SocketServer.kt:
 (96, 33): Overload resolution ambiguity:
public fun PokemonDataOuterClass.PokemonData.getIvPercentage(): Int defined in i
nk.abb.pogo.scraper.util.pokemon
public fun PokemonDataOuterClass.PokemonData.getIvPercentage(): Int defined in i
nk.abb.pogo.scraper.util.pokemon
e: C:\Dev\PokemonGoBot\src\main\kotlin\ink\abb\pogo\scraper\tasks\CatchOneNearby
Pokemon.kt: (38, 129): Overload resolution ambiguity:
public fun PokemonDataOuterClass.PokemonData.getIvPercentage(): Int defined in i
nk.abb.pogo.scraper.util.pokemon
public fun PokemonDataOuterClass.PokemonData.getIvPercentage(): Int defined in i
nk.abb.pogo.scraper.util.pokemon
e: C:\Dev\PokemonGoBot\src\main\kotlin\ink\abb\pogo\scraper\util\pokemon\Pokemon
.kt: (26, 1): 'public fun PokemonDataOuterClass.PokemonData.getIv(): Int' confli
cts with another declaration in package 'pokemon'
e: C:\Dev\PokemonGoBot\src\main\kotlin\ink\abb\pogo\scraper\util\pokemon\Pokemon
.kt: (31, 1): 'public fun PokemonDataOuterClass.PokemonData.getIvPercentage(): I
nt' conflicts with another declaration in package 'pokemon'
e: C:\Dev\PokemonGoBot\src\main\kotlin\ink\abb\pogo\scraper\util\pokemon\Pokemon
.kt: (32, 14): Overload resolution ambiguity:
public fun PokemonDataOuterClass.PokemonData.getIv(): Int defined in ink.abb.pog
o.scraper.util.pokemon
public fun PokemonDataOuterClass.PokemonData.getIv(): Int defined in ink.abb.pog
o.scraper.util.pokemon
e: C:\Dev\PokemonGoBot\src\main\kotlin\ink\abb\pogo\scraper\util\pokemon\Pokemon
Data.kt: (13, 1): 'public fun PokemonDataOuterClass.PokemonData.getIv(): Int' co
nflicts with another declaration in package 'pokemon'
e: C:\Dev\PokemonGoBot\src\main\kotlin\ink\abb\pogo\scraper\util\pokemon\Pokemon
Data.kt: (18, 1): 'public fun PokemonDataOuterClass.PokemonData.getIvPercentage(
): Int' conflicts with another declaration in package 'pokemon'
e: C:\Dev\PokemonGoBot\src\main\kotlin\ink\abb\pogo\scraper\util\pokemon\Pokemon
Data.kt: (19, 14): Overload resolution ambiguity:
public fun PokemonDataOuterClass.PokemonData.getIv(): Int defined in ink.abb.pog
o.scraper.util.pokemon
public fun PokemonDataOuterClass.PokemonData.getIv(): Int defined in ink.abb.pog
o.scraper.util.pokemon
:compileKotlin FAILED`

**Comment by iamthasasquatch** [comment]
alright got it working few minor issues I am having. The "Log" window(mimic of console) doesnt auto scroll down when new data is added. Notifications anyway we can save the pokemon caught list into that window instead of notifications kind of like a list saying rattata was caught, pidgey was caught etc or have notification but the stored list as well. Also refreshing the page or reloading it doesnt seem to save the window positions where i moved them.

**Comment by Simpuhl** [comment]
This is looking amazing, can't wait to see a finished release we can all try

**Comment by Emuuung** [comment]
Fix branch conflicts so this gets right at the top of the list :+1:

**Comment by Emuuung** [comment]
@shilch It works great, I just can't resize any of the windows, am I missing a trick to it? cause u try placing mouse on borders to resize but nothing happens.
Also the log should have an "auto-scroll down" property because it looks like it's still.

**Comment by pashorizer** [comment]
@jabbink Do you plan on making this feature non-optional?

**Comment by Emuuung** [comment]
@pashorizer the `config.properties.template` file has the option to select port (default `8000`) and set to `0` to disable it.

**Comment by angrox** [comment]
Is the GUI capable of running on another server? When I check the html code it always wants a connect to localhost:8001?

**Comment by shilch** [comment]
Yes, please edit the port in the html. I did not implement the option to change the socket port

**Comment by shilch** [comment]
Now you can right click on the map to create a route that the player should follow ( #161  )

**Comment by Emuuung** [comment]
This PR looks better and better every day.
But I just tired it with the latest commits from develop branch and bot gets stuck in the same pokestop.
Also now I can resize all the windows, but some windows should be resize-able to smaller widths/heights if user wants to.

**Comment by shilch** [comment]
GPX upload support is included now ( @victorespada ).
I need some help with the eggs window. Which function gives me the walked kms of an egg?

**Comment by shilch** [comment]
Ok, showing km of an egg will be available in a newer version of the PokemonGO Api.
So you can merge this PR.

**Comment by alone23** [comment]
Really nice, I'm exited :)

File: src/main/kotlin/ink/abb/pogo/scraper/gui/SocketServer.kt [hunk 1]
```diff
@@ -0,0 +1,218 @@
...
+import com.corundumstudio.socketio.SocketIOServer
+import com.pokegoapi.api.map.fort.Pokestop
+import com.pokegoapi.api.player.PlayerProfile
+import com.pokegoapi.api.pokemon.Pokemon
+import com.pokegoapi.google.common.geometry.S2LatLng
+import ink.abb.pogo.scraper.Context
+import ink.abb.pogo.scraper.util.pokemon.getIvPercentage
+
+class SocketServer {
+    val requiredXp = arrayOf(0, 1000, 3000, 6000, 10000, 15000, 21000, 28000, 36000, 45000, 55000, 65000, 75000,
```
Comment (jabbink) [inline src/main/kotlin/ink/abb/pogo/scraper/gui/SocketServer.kt]: Now we have the same magic number array twice in the code, maybe refactor so it's only defined once

File: src/main/kotlin/ink/abb/pogo/scraper/util/pokemon/PokemonData.kt [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+ * This is free software, and you are welcome to redistribute it under certain conditions.
+ *
+ * For more information, refer to the LICENSE file in this repositories root directory
+ */
+
+package ink.abb.pogo.scraper.util.pokemon
+
+import POGOProtos.Data.PokemonDataOuterClass
+
+fun PokemonDataOuterClass.PokemonData.getIv(): Int {
```
Comment (jabbink) [inline src/main/kotlin/ink/abb/pogo/scraper/util/pokemon/PokemonData.kt]: already defined in ink/abb/pogo/scraper/util/pokemon/Pokemon.kt:24 as well, maybe remove it there.

**Comment by Emuuung** [comment]
Is there a way to center map so the map windows always follow your character?

**Comment by shilch** [comment]
@jabbink Yes I will refactor this array, this extension of class PokemonData is necessary and can not be replaced with the Pokemon.kt

@Emuuung I disabled it because you may like to move the map and the map would always center back the player. Maybe I should put an "autofollow" checkbox right to the GPX button

**Comment by Emuuung** [comment]
@shilch Ah I see, to move the map the user should move it freely, I guess your idea about adding an option to autofolllow is much better :wink:

**Comment by iamthasasquatch** [comment]
the error from eggs is because its not in yet right? do i need to comment out that section? 

e: C:\Dev\PokemonGoBot\src\main\kotlin\ink\abb\pogo\scraper\gui\SocketServer.kt:
 (137, 45): Unresolved reference: eggKmWalkedStart
:compileKotlin FAILED

**Comment by shilch** [comment]
Mmh, I do not get this error but yes, you can comment out the eggs section in SocketServer.kt

**Comment by iamthasasquatch** [comment]
any idea why i would be getting that error then? there is only a few callings to that

**Comment by shilch** [comment]
Probably your cloned project includes a newer Version of the API (git submodule) I think `eggKmWalkedStart` was removed today

**Comment by iamthasasquatch** [comment]
possibly i did run it shortly before i had posted I always run it when doing new build ;/

**Comment by Emuuung** [comment]
@shilch map autofollow feature is working great! Nice work :+1:

**Comment by iamthasasquatch** [comment]
mine shows 13/350 as well lol
I had to comment out the egg stuff since my api was updated and it was removed.
I also edited the zoom to 14 on the index.html, and made my autofollow to default to on works great besides the item count being off

**Comment by shilch** [comment]
For me it shows 14 / 350 items, it seems to be a bug of the api. Ignore this first. @pr0ves wrote some code that should fix this, I will try it.
@iamthasasquatch the pokebank gets refresh every time a new pokemon gets added to the list. You have to scroll to bottom to see it. They are not sorted yet.

File: src/main/kotlin/ink/abb/pogo/scraper/gui/SocketServer.kt [hunk 3]
```diff
@@ -0,0 +1,215 @@
...
+            profile.stardust = ctx!!.api.playerProfile.currencies[PlayerProfile.Currency.STARDUST]
+            profile.level = ctx!!.api.playerProfile.stats.level
+            val curLevelXP = ctx!!.api.playerProfile.stats.experience - requiredXp[ctx!!.api.playerProfile.stats.level - 1]
+            profile.levelXp = curLevelXP
+            val nextXP = requiredXp[ctx!!.api.playerProfile.stats.level] - requiredXp[ctx!!.api.playerProfile.stats.level - 1]
+            val ratio = ((curLevelXP.toDouble() / nextXP.toDouble()) * 100).toInt()
+            profile.levelRatio = ratio
+            profile.pokebank = ctx!!.api.inventories.pokebank.pokemons.size
+            profile.pokebankMax = ctx!!.api.playerProfile.pokemonStorage
+            profile.items = ctx!!.api.inventories.itemBag.items.size
```
Comment (pr0ves) [inline src/main/kotlin/ink/abb/pogo/scraper/gui/SocketServer.kt]: This gives you the number of item types, use itemBag.size()

**Comment by shilch** [comment]
@pr0ves Ahh, thank you. That explains the low number. But how to get the maximal size. 350 is incorrect, right?

**Comment by pr0ves** [comment]
It's correct if you haven't bought an upgrade for the bag, I don't know if it works with the upgrades.

**Comment by shilch** [comment]
I think the bag size is 250? I only bought an pokebank upgrade

File: src/main/kotlin/ink/abb/pogo/scraper/gui/SocketServer.kt [hunk 4]
```diff
@@ -63,7 +63,14 @@ class SocketServer {
...
             profile.pokebankMax = ctx!!.api.playerProfile.pokemonStorage
-            profile.items = ctx!!.api.inventories.itemBag.items.size
+
+            // Replace with @pr0ves method
+            var count = 0
+            ctx!!.api.inventories.itemBag.items.toMutableList().forEach {
+                count += it.count
+            }
+            profile.items = count
+
```
Comment (pr0ves) [inline src/main/kotlin/ink/abb/pogo/scraper/gui/SocketServer.kt]: You can just use `ctx!!.api.inventories.itemBag.size()` directly, without copypasting my function here.

**Comment by shilch** [comment]
OK, changed it

**Comment by Emuuung** [comment]
Oh wow, so many changes, can't wait to test this, sadly servers are down :disappointed:

**Comment by shilch** [comment]
My bot is running without network errors

**Comment by Emuuung** [comment]
@shilch oh, it's only down for PTC accounts.

**Comment by iamthasasquatch** [comment]
is the eggs fixed in this or still working on it?

**Comment by shilch** [comment]
WIP, will open a second PR

**Comment by jabbink** [comment]
second PR isn't necessary afaik

**Comment by Emuuung** [comment]
The pokebank window sort pokemon by CP, is there anyway to sort it by IV? Might be better for people using the `sort_by_iv=true` setting in `config.properties`.

**Comment by JPhix** [comment]
27 Jul 18:05:59: Running webserver on port 8000
SLF4J: Failed to load class "org.slf4j.impl.StaticLoggerBinder".
SLF4J: Defaulting to no-operation (NOP) logger implementation
SLF4J: See http://www.slf4j.org/codes.html#StaticLoggerBinder for further details.

i seem to be getting the same error on both linux and windows 
linux the webserver runs fine but socket server doesnt seem to be listening
java version "1.8.0_101"
Java(TM) SE Runtime Environment (build 1.8.0_101-b13)
Java HotSpot(TM) 64-Bit Server VM (build 25.101-b13, mixed mode)
debian 8.4 (Jessie)
and then on my windows machine it says its listening on port 8000 but noting loads in the browser but  8001 does load a blank page

**Comment by ArtursGailis1995** [comment]
On Windows 8.1 WEB GUI was working very well. Tested on 2 Windows 10 personal machines and on localhost:8000 I get "Page not working" error and on localhost:8001 I get a blank page. Allowed in firewall both times, tried disabling firewall and antivirus, tried rebuilding. Tried reinstalling all Java installations on both machines. Nothing helps to make it work on Windows 10. Also tried changing ports and running on different browsers. No other software is using local WEB server at the moment.

**Comment by shilch** [comment]
@ArtursGailis1995 Interesting, which java version do you have on windows 8.1 and the two windows 10 machines (version and jre / jdk) ?

**Comment by ArtursGailis1995** [comment]
@shilch Its the same config on all these machines. I think it is something to do with how Windows 10 treats DNS caching because internet is 'full' of similar issues. Tried all possible solutions found on internet, but it does not work at all on W10.
untitled

**Comment by jabbink** [comment]
Can you give the output of `java -version`? (you have both 64 and 32 bit installed, who knows which one it picks)

**Comment by karael** [comment]
same with java 8 update 102

**Comment by ArtursGailis1995** [comment]
@jabbink I have attached it below. I have all these versions installed on all three computers. And first time after installing Java it pops up firewall windo where I allow it to access network (it is trying to work). Firefox says 'connection reset'. Chrome gives 'ERR_EMPTY_RESPONSE'.
untitled2

**Comment by Emuuung** [comment]
@ArtursGailis1995 Try this [java jdk](http://www.oracle.com/technetwork/java/javase/downloads/jdk8-downloads-2133151.html), it might be because you don't have the Development Kit.

**Comment by JPhix** [comment]
made no difference for me installing the java jdk as well on windows 10 
java version "1.8.0_101"
Java(TM) SE Runtime Environment (build 1.8.0_101-b13)
Java HotSpot(TM) 64-Bit Server VM (build 25.101-b13, mixed mode)

**Comment by ArtursGailis1995** [comment]
@Emuuung @JPhix @jabbink @shilch I was hoping to contribute to better WEB GUI but cant do it because Windows 10 does not support it. Cant really develop on my work (Windows 8.1) computer.

**Comment by Emuuung** [comment]
mine is:
```
java version "1.8.0_102"
Java(TM) SE Runtime Environment (build 1.8.0_102-b14)
Java HotSpot(TM) 64-Bit Server VM (build 25.102-b14, mixed mode)
```
Maybe this one makes a difference?

**Comment by nullsumme** [comment]
hmm would it be possible to integrate pokevision somehow? would be perfekt for farming special pokemons

**Comment by JPhix** [comment]
@Emuuung are you running windows 10 though ?

**Comment by jabbink** [comment]
Are you guys trying http://127.0.0.1:8000 or http://localhost:8000

**Comment by jabbink** [comment]
And please continue the discussion on #493

**Comment by JPhix** [comment]
both have even tryed my local lan ip address @jabbink

File: src/main/kotlin/ink/abb/pogo/scraper/gui/WebServer.kt [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+            server.executor = null
+            server.start()
+        }.start()
+    }
+
+    inner class RootHandler(val socketPort: Int) : HttpHandler {
+
+        @Throws(IOException::class)
+        override fun handle(t: HttpExchange) {
+            var string = String(Files.readAllBytes(Paths.get("src/main/resources/gui/index.html")))
```
Comment (GBakatoS) [inline src/main/kotlin/ink/abb/pogo/scraper/gui/WebServer.kt]: Path to HTML file is not correct when running from JAR

Comment (jabbink) [inline src/main/kotlin/ink/abb/pogo/scraper/gui/WebServer.kt]: Thanks, should be fixed to getResource(Stream)

**Comment by swtrpeter** [comment]
I need some help here.. I load the webpage and there is no map and it does not look anything like yours. I followed the steps as i should but it isnt working. Please help me out

**Comment by Rendiere** [comment]
@swtrpeter Try dragging the windows that you see around. The default starting page contains all the windows (log, profile, map) on top of each other. 

@shilch  @iamthasasquatch Just want to give a big THANK YOU YOU GUYS ARE AMAZING!!!

**Pull request merged**